### PR TITLE
Add parameter required for Google Maps Geocoding API.

### PR DESCRIPTION
**Pull request by chris001**
The JJW Design Google Maps was using the old Google Maps v3 "hash" parameter.
The API has changed to Google Maps Geocoding API and requires a "key" parameter instead.

NOTE the Geocoding API is used by the server through cron jobs to periodically obtain geographic coordinates (latitude, longitude) of physical street addresses, for the purposes of later displaying pins on the geographical maps in the SutieCRM user's web browser.  

Google Maps Geocoding API is different from the Google Maps Javascript API which is used to DISPLAY the geographical maps on the SuiteCRM user's web browser.  

Two separate keys, for two separate uses, and two separate daily quotas for free usage.

## Description

Add "key" parameter to URL, corresponding to the "secret" field on the Google Maps module's config page.
Add a convenient URL on the Google Maps module's config page to make it easy for user to get a free Google Maps Geocoding API key.

## Motivation and Context

The Google Maps module's usage of Geocoding API has been failing on new IP addresses (new servers) since June 2016.
The parameters have changed.  Old parameter = "hash" = md5 hash of address concatenated with Goole Maps v3 API key.  New parameter = "key" = plaintext Google Maps Geocoding API key.
More details:
https://developers.google.com/maps/documentation/geocoding/get-api-key#specify-a-key-in-your-request

## How To Test This

1. Install SuiteCRM on a new IP address which has never before used SuiteCRM's JJWG_maps module to Geocode addresses.
2. Try to geocode addresses.  It should fail.
3. Get an API key by going to Admin, Google Maps Config, Get a free API key, Create project, get Geocoding API key, copy/paste it into the SuiteCRM Google Maps Config page.
4. Try to geocode addresses.  It should succeed.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: modules/jjwg_Maps/views/view.config.php [hunk 1]
```diff
@@ -423,6 +423,10 @@ function display() {
             title='' tabindex='140' size="25" maxlength="255">
             &nbsp; <?php echo $GLOBALS['mod_strings']['LBL_CONFIG_DEFAULT']; ?>
                 <?php echo htmlspecialchars($GLOBALS['jjwg_config_defaults']['geocoding_api_secret']) ?>
+            <br/>
+            <a href='https://developers.google.com/maps/documentation/geocoding/get-api-key' target='_blank'>
```
Comment (horus68) [inline modules/jjwg_Maps/views/view.config.php]: Can you configure this to enable translations? 
See related PR https://github.com/salesagility/SuiteCRM/pull/2536

Comment (chris001) [inline modules/jjwg_Maps/views/view.config.php]: OK you're right I'll add string labels in place of hard coded english.

File: modules/jjwg_Maps/language/en_us.lang.php [hunk 2]
```diff
@@ -143,6 +143,7 @@
 $mod_strings['LBL_CONFIG_GEOCODING_API_URL_DESC'] = 'The URL of the Google Maps API V3 or Proxy';
 $mod_strings['LBL_CONFIG_GEOCODING_API_SECRET_TITLE'] = 'Secret Phrase for Proxy:';
 $mod_strings['LBL_CONFIG_GEOCODING_API_SECRET_DESC'] = 'The Secret Phrase to be used with the Proxy MD5 comparison.';
+$mod_strings['LBL_CONFIG_GEOCODING_API_SECRET_GET_A_KEY'] = 'Get a free Google Maps Geocoding API key (to geocode 2500 addresses per day for free).';
```
Comment (horus68) [inline modules/jjwg_Maps/language/en_us.lang.php]: Better not include the free limits details `(to geocode 2500 addresses per day for free)` as this can be changed by google in the future and it will require new translations.

File: modules/Campaigns/SubPanelViewer.php [hunk 3]
```diff
@@ -87,7 +87,7 @@
 
 $subpanel_object = new SubPanel($module, $record, $subpanel,null, $layout_def_key);
 
-$subpanel_object->setTemplateFile('include/SubPanel/SubPanelDynamic.html');
```
Comment (samus-aran) [inline modules/Campaigns/SubPanelViewer.php]: Is this related or not?

**Comment by Dillon-Brown** [comment]
Hi @chris001 Could you clarify where exactly the geocode addresses should fail? It seems to be returning the same with/without your fix.

File: modules/Campaigns/SubPanelViewer.php (hunk 3 above)
Comment (chris001) [inline modules/Campaigns/SubPanelViewer.php]: No, this isn't related. Will have to remove it.

**Comment by chris001** [comment]
@Dillon-Brown   The geocode API will fail under the following circumstances:
* Your server should run on an IP address which has NOT previously been whitelisted by Google in mid-2016 or before, to use the free Geocoding API without an API key, for unlimited "reasonable use".  During the year 2016 is when their policy on unlimited reasonable free use changed.  API key is required, and there is now a limited cap on free use, 2500 addresses per day.

**Comment by horus68** [comment]
@Dillon-Brown Is this PR still relevant? 
This has to be confirmed by people using a new site IP (or site created after 2016)

**Comment by markbond1007** [comment]
Yes this is still relevant and can be fixed easily (I have just tested this). On a brand new IP I ran geocode (with a valid Google Key), not all data was geocoded. After this using the quick Test Geocode on an address tells me I am over quota (on an account that has billing enabled etc. at google).

To fix you need to go to modules/jjwg_Maps/jjwg_Maps.php go to line 993 (ish) in the function getGoogleMapsGeocode

replace:
`if (!empty($this->settings['geocoding_api_secret'])) {
            $hash = md5($address.$this->settings['geocoding_api_secret']);
            $request_url .= '&hash='.urlencode($hash);
        }`

with:

`$request_url.=&key=`.urlencode($this->settings['google_maps_api_key']);

**Comment by horus68** [comment]
@Dillon-Brown can you create a new PR for this?

**Comment by pgorod** [comment]
If this "needs updates", what are the updates it needs?

**Comment by markbond1007** [comment]
wasn't this fixed in another PR? I have been using this regularly with the previous patches provided and all seems okay!

**Comment by pgorod** [comment]
Ok, so let's close! 
@samus-aran @Dillon-Brown
